File: lab5/main.cpp
#include <iostream>
using namespace std;

// Question 1: factorial (recursive)
int factorial(int x) {
    if (x == 0 || x == 1)
        return 1;
    else
        return x * factorial(x - 1);
}


// Question 2: e^x calculation
double exp_x(int x, int n) {
    double total = 0;

    for (int i = 0; i <= n; i++) {
        // term = (x^i) / i!

        // x^i hesapla
        double power = 1;
        for (int j = 0; j < i; j++) {
            power = power * x;
        }

        double term = power / factorial(i);
        total = total + term;
    }

    return total;
}


// Question 3 & 4: alternating series (recursive, return var)
double series(int n) {
    if (n == 1)
        return 1;

    if (n % 2 == 0)
        return series(n - 1) - (1.0 / n);
    else
        return series(n - 1) + (1.0 / n);
}


// MAIN
int main() {

    // 1
    int num;
    cout << "Enter number for factorial: ";
    cin >> num;
    cout << "Factorial = " << factorial(num) << endl;

    cout << endl;

    // 2
    int x, n;
    cout << "Enter x: ";
    cin >> x;
    cout << "Enter n: ";
    cin >> n;

    cout << "e^x approximation = " << exp_x(x, n) << endl;

    cout << endl;

    // 3 & 4
    int n2;
    cout << "Enter n for series: ";
    cin >> n2;

    cout << "Series result = " << series(n2) << endl;

    return 0;
}

File: lab5/main.py
# Question 1
def factorial(x):
    if x == 0 or x == 1:
        return 1
    else:
        return x * factorial(x - 1)


# Question 2
term = lambda x, n: (x ** n) / factorial(n)

def exp_x(x, n):
    total = 0
    i = 0
    while i <= n:
        total = total + term(x, i)
        i = i + 1
    return total


# Question 3
total_sum = 0

def alternating_series(n):

    global total_sum

    if n == 0:
        return

    alternating_series(n - 1)

    if n % 2 == 1:
        total_sum = total_sum + (1 / n)
    else:
        total_sum = total_sum - (1 / n)


# MAIN PART
print("1. Factorial")
num = int(input("Enter a number for factorial: "))
print("Factorial =", factorial(num))

print()

print("2. e^x calculation")
x = int(input("Enter x: "))
n = int(input("Enter n: "))
print("Result =", exp_x(x, n))

print()

print("3. Alternating series")
n2 = int(input("Enter n: "))
alternating_series(n2)
print("Series result =", total_sum)